framework/src/modules/pos/module.ts

```diff
@@ -755,7 +756,16 @@ export class PoSModule extends BaseModule {
 			const numInitValidators =
 				genesisData.initRounds + this._moduleConfig.numberActiveValidators - round;
 			const numElectedValidators = this._moduleConfig.numberActiveValidators - numInitValidators;
-			const activeValidators = snapshotValidators.slice(0, numElectedValidators);
+			let weightSum = BigInt(0);
+			const activeValidators = snapshotValidators.slice(0, numElectedValidators).map(v => {
+				const scaledWeight = this._ceiling(v.weight, WEIGHT_SCALE_FACTOR);
+				weightSum += scaledWeight;
+				return {
+					...v,
+					weight: scaledWeight,
+				};
+			});
+			const averageWeight = weightSum / BigInt(numElectedValidators);
 			for (const address of genesisData.initValidators) {
```

Comment:
Comment about line 601 (and `snapshotStore` more generally): There is a mis match between specs and implementation on the snapshotStore. At the end of round `r`, according to LIP we set the value `snapshotStore(r)` and delete `snapshotStore(r-3)` while here in the implementation you set snapshotStore(r+3) and delete `snapshotStore(r)`. 

Although it is kind of equivalent, I think this is a deviation from protocol; in particular the state root you get is different than the one described in the protocol.

Reply:
New code seems correct, can be resolved from my side.